### PR TITLE
fix(#7007): Implements Vault Health Check integration.

### LINKED ISSUE #7007: Add health check for Vault
**Description**

Add health check for Vault.

**Implementation ideas**

Create a readiness health check that checks if Vault is up and running.

(I can provide a PR as I have implemented for my project)

### PR DESCRIPTION
**Pull request by lordofthejars**
Fixes #7007

### CODE REVIEW

File: extensions/vault/runtime/src/main/java/io/quarkus/vault/health/VaultHealthCheck.java
Comment (sberyozkin) [inline extensions/vault/runtime/src/main/java/io/quarkus/vault/health/VaultHealthCheck.java]: The standby and perf customizations are not possible here

File: extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/VaultSystemBackendManager.java
Comment (sberyozkin) [inline extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/VaultSystemBackendManager.java]: Would you like to keep these parameters to let the user code override the configuration properties ? I propose just drop them since the config properties win anyway and the user code does not probably need to know about these minor status tweaks

File: extensions/vault/runtime/src/main/java/io/quarkus/vault/health/VaultHealthCheck.java
Comment (lordofthejars) [reply extensions/vault/runtime/src/main/java/io/quarkus/vault/health/VaultHealthCheck.java]: Not here, but if they are configured in `application.properties`, then these values are overridden. I know that you are asking why are you leaving this instead of provide a `health()` method. The reason is that `VaultSystemBackendEngine` could be injected in any class and users programmatically might want to set them programmatically.

File: extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/VaultSystemBackendManager.java
Comment (lordofthejars) [reply extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/VaultSystemBackendManager.java]: I copy the same comment before just to keep a relationship:

I know that you are asking why are you leaving this instead of provide a health() method. The reason is that VaultSystemBackendEngine could be injected in any class and users programmatically might want to set them programmatically.

File: extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/VaultRuntimeConfig.java
Comment (sberyozkin) [inline extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/VaultRuntimeConfig.java]: I propose to introduce a `Health` configuration group as these 3 parameters really stand up on their own IMHO.
Also, is default value needed for `false` ?

Comment (lordofthejars) [reply extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/VaultRuntimeConfig.java]: Now, correct me if I am not right these properties are under: `quarkus.vault.health.*` isn't it?

Yes by default they are `false`

**Comment by sberyozkin** [comment]
@lordofthejars Hi, it looks like a very useful addition, lets see what Vincent @vsevel says

File: extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/VaultRuntimeConfig.java
Comment (lordofthejars) [reply extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/VaultRuntimeConfig.java]: What I have seen in other extensions that they are integrating with the spec, they are following the same approach, but please I am new on extension ecosystem, so I can be totally wrong.

File: extensions/vault/runtime/src/main/java/io/quarkus/vault/health/VaultHealthCheck.java
Comment (sberyozkin) [reply extensions/vault/runtime/src/main/java/io/quarkus/vault/health/VaultHealthCheck.java]: @lordofthejars thanks, but then we'd support the users overwriting the configuration, and when the config is good, then they'd have to type `false, false` :-)
Having `health()` alone is quite enough IMHO, it does seem to me in most cases health of the running active Vault would be of interest. `health(boolean standbyok)` and `health(boolean perfok)` can be easily added if someone ever asks :-)

File: extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/VaultRuntimeConfig.java
Comment (sberyozkin) [reply extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/VaultRuntimeConfig.java]: You'd have 
```
@ConfigItem
Health health = new Health();
@ConfigGroup
public class Health {
   boolean enabled;
   // etc
}
```

Sure lets keep a default initialization to `false`

Comment (sberyozkin) [reply extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/VaultRuntimeConfig.java]: @lordofthejars It is not a big deal, but it feels it would better be encapsulated in the dedicate group

**Comment by sberyozkin** [comment]
@lordofthejars or may be have 2 health() methods from the beginning :-), a no arg method to make a mainstream case easier to support, and with 2 booleans for the customizations

**Comment by lordofthejars** [comment]
@sberyozkin there some failures in CI/CD but running from Maven from my machine they are not happening.

The failure is:

```
2020-03-16T17:59:59.9522373Z [INFO] Running io.quarkus.vault.VaultITCase
2020-03-16T18:00:00.1382299Z [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.177 s <<< FAILURE! - in io.quarkus.vault.VaultITCase
2020-03-16T18:00:00.1388461Z [ERROR] io.quarkus.vault.VaultITCase  Time elapsed: 0.177 s  <<< ERROR!
2020-03-16T18:00:00.1398329Z java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: Failed to load steps from class io.quarkus.vault.VaultProcessor
2020-03-16T18:00:00.1404703Z 	at io.quarkus.test.QuarkusUnitTest.beforeAll(QuarkusUnitTest.java:427)
2020-03-16T18:00:00.1413271Z 	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllCallbacks$7(ClassBasedTestDescriptor.java:359)
2020-03-16T18:00:00.1421923Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-03-16T18:00:00.1429960Z 	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllCallbacks(ClassBasedTestDescriptor.java:359)
2020-03-16T18:00:00.1437115Z 	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:189)
2020-03-16T18:00:00.1445676Z 	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:78)
2020-03-16T18:00:00.1453922Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:132)
2020-03-16T18:00:00.1465664Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-03-16T18:00:00.1496795Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
2020-03-16T18:00:00.1502335Z 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
2020-03-16T18:00:00.1509810Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
2020-03-16T18:00:00.1517231Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-03-16T18:00:00.1540396Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
2020-03-16T18:00:00.1548070Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
2020-03-16T18:00:00.1551781Z 	at java.util.ArrayList.forEach(ArrayList.java:1257)
2020-03-16T18:00:00.1562483Z 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
2020-03-16T18:00:00.1578429Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
2020-03-16T18:00:00.1585486Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-03-16T18:00:00.1600541Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
2020-03-16T18:00:00.1622399Z 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
2020-03-16T18:00:00.1665052Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
2020-03-16T18:00:00.1672381Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-03-16T18:00:00.1679340Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
2020-03-16T18:00:00.1685341Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
2020-03-16T18:00:00.1700172Z 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
2020-03-16T18:00:00.1732895Z 	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
2020-03-16T18:00:00.1741209Z 	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
2020-03-16T18:00:00.1747863Z 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:220)
2020-03-16T18:00:00.1774736Z 	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:188)
2020-03-16T18:00:00.1781968Z 	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:202)
2020-03-16T18:00:00.1788204Z 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:181)
2020-03-16T18:00:00.1796125Z 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:128)
2020-03-16T18:00:00.1797215Z 	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:142)
2020-03-16T18:00:00.1798077Z 	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:117)
2020-03-16T18:00:00.1824594Z 	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
2020-03-16T18:00:00.1830485Z 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
2020-03-16T18:00:00.1841250Z 	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
2020-03-16T18:00:00.1865547Z 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
2020-03-16T18:00:00.1888285Z Caused by: java.lang.RuntimeException: java.lang.RuntimeException: Failed to load steps from class io.quarkus.vault.VaultProcessor
2020-03-16T18:00:00.1896800Z 	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:206)
2020-03-16T18:00:00.1943324Z 	at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:95)
2020-03-16T18:00:00.1948429Z 	at io.quarkus.test.QuarkusUnitTest.beforeAll(QuarkusUnitTest.java:386)
2020-03-16T18:00:00.1950047Z 	... 37 more
2020-03-16T18:00:00.1997638Z Caused by: java.lang.RuntimeException: Failed to load steps from class io.quarkus.vault.VaultProcessor
2020-03-16T18:00:00.1998611Z 	at io.quarkus.deployment.ExtensionLoader.loadStepsFrom(ExtensionLoader.java:214)
2020-03-16T18:00:00.1999447Z 	at io.quarkus.deployment.ExtensionLoader.loadStepsFrom(ExtensionLoader.java:136)
2020-03-16T18:00:00.2000273Z 	at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:101)
2020-03-16T18:00:00.2001105Z 	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:204)
2020-03-16T18:00:00.2001941Z 	... 39 more
2020-03-16T18:00:00.2002826Z Caused by: java.lang.IllegalArgumentException: Run time configuration cannot be consumed here unless the method is a @Recorder at io.quarkus.vault.runtime.config.VaultRuntimeConfig arg0 of io.quarkus.smallrye.health.deployment.spi.HealthBuildItem io.quarkus.vault.VaultProcessor.addHealthCheck(io.quarkus.vault.runtime.config.VaultRuntimeConfig) of class io.quarkus.vault.VaultProcessor
2020-03-16T18:00:00.2003722Z 	at io.quarkus.deployment.ExtensionLoader.reportError(ExtensionLoader.java:992)
2020-03-16T18:00:00.2008270Z 	at io.quarkus.deployment.ExtensionLoader.loadStepsFrom(ExtensionLoader.java:748)
2020-03-16T18:00:00.2013463Z 	at io.quarkus.deployment.ExtensionLoader.loadStepsFrom(ExtensionLoader.java:212)
2020-03-16T18:00:00.2015075Z 	... 42 more
```

Since I cannot reproduce locally and before I start doing pushes to test it, does it mean that I need to do:

```
@ConfigGroup
public class HealthConfig {

    public static final String DEFAULT_HEALTH_CHECK_ENABLED = "false";

    /**
     * Whether or not an health check is published in case the smallrye-health extension is present.
     */
    @Recorder
    @ConfigItem(name = "enabled", defaultValue = DEFAULT_HEALTH_CHECK_ENABLED)
    public boolean enabled;
}
```

**Comment by geoand** [comment]
Hi @lordofthejars. 

What you are showing in the code snippet is not correct.
The error essentially tells you that:

```
    HealthBuildItem addHealthCheck(VaultRuntimeConfig config) {
        return new HealthBuildItem("io.quarkus.vault.health.VaultHealthCheck",
                config.healthEnabled, "vault");
    }
```

cannot consume `VaultRuntimeConfig`. This is because at build time (when `addHealthCheck` runs), the runtime value of `config` is not known.
The only legal operation that is permitted to do on a runtime config value, is to pass it to a recorder class (see https://quarkus.io/guides/writing-extensions#bytecode-recording for details on a  recorder).

I would suggest that you do a

```
./mvnw clean install -DskipTests -DskipITs -DskipDocs
```

and then run the tests locally. It is almost certain that you will be able to reproduce the problem you are seeing in CI.

**Comment by lordofthejars** [comment]
@geoand So I need to annotate the Config class as `@Recorder`

**Comment by geoand** [comment]
> @geoand So I need to annotate the Config class as `@Recorder`

That is not the correct use of `@Recorder`. See the extension author's guide I linked to

**Comment by lordofthejars** [comment]
From Stuart conversation:

This is runtime config, but you are reading it at build time, so runtime changes would not be picked up.

You need to create a new config class that is build time config, see io.quarkus.vertx.http.runtime.HttpConfiguration and io.quarkus.vertx.http.runtime.HttpBuildTimeConfig for an example.

**Comment by geoand** [comment]
> From Stuart conversation:
> 
> This is runtime config, but you are reading it at build time, so runtime changes would not be picked up.
> 
> You need to create a new config class that is build time config, see io.quarkus.vertx.http.runtime.HttpConfiguration and io.quarkus.vertx.http.runtime.HttpBuildTimeConfig for an example.

To expand on this: Build time configuration can be read without issues in build step. That is why Stuart proposes to split the vault configuration into build time and runtime (as other extensions do) and add the health configuration as a build time property.

**Comment by lordofthejars** [comment]
I thought that this could happen when I was developing it but I was not super sure. Well, every day you learn something new.

**Comment by lordofthejars** [comment]
@vsevel @geoand @stuartwdouglas The only failing it seems it is not related to my change as only fails in windows and in RestEasy module. The rest of the checks are successful.

**Comment by sberyozkin** [comment]
@lordofthejars IMHO it is a good idea to introduce `VaultBuildTimeConfig` but it should probably have an `enable` property only which enables/disables all the Vault extension.
I'm also not sure a `Health` config group is needed because the standby and perf boolean properties have gone, I've only proposed to have a group because there were 3 health related properties :-).

I wonder, is `health.enabled` property needed at all ? Vault `enable` property should be enough, if Vault is enabled (which is default) then letting the users check its status is reasonable ? This Vault `health` path can be protected if needed. CC @vsevel

**Comment by lordofthejars** [comment]
@sberyozkin As I understood, about these two properties we decided to drop until someone requests them. About the enable flag it was discussed here: https://github.com/quarkusio/quarkus/issues/7007#issuecomment-583395622

**Comment by sberyozkin** [comment]
@lordofthejars OK, if it is preferred to have it optionally enabled, then it is fine. 

Re those 2 extra health properties. Since they are gone, `Health` group is not needed, however may be lets keep it if Vincent will prefer to have them configured.
What I was saying earlier, is that from the user code point's of view, typing `health(false, false)` for the mainstream case seemed too much. I'd rather just leave `health()` only and let users tweak how Vault responds with those 2 extra booleans in the `Health` group. Yes, I also typed the proposal to have `health(boolean, boolean)` :-), but if Vincent supports configuring those 2 props too then it would become a bit messy with describing what wins in `health(boolean, boolean)` vs the configuration. Not a big deal, but please think if it is reasonable...
Thanks.
Overall looks as a very useful addition IMHO

**Comment by lordofthejars** [comment]
@sberyozkin Ok, tomorrow morning I'll wake up a bit earlier and I'll send the PR so you can see it and review.

**Comment by sberyozkin** [comment]
Hey @lordofthejars look forward to it and I'm sorry for sending the confusing messages earlier :-) thanks

**Comment by lordofthejars** [comment]
@sberyozkin  No worries, changes done.

**Comment by sberyozkin** [comment]
Hi @lordofthejars Thanks, IMHO it is nearly ready to go, pending an approval from @vsevel. Also CC @gsmet in case Guillaume may have some comments too. Oh yeah, can you squash please :-) ?
May be also consider dropping `ConfigItem` `name` property for all the health properties. 
Thanks

**Comment by lordofthejars** [comment]
@sberyozkin Squashed and name property dropped down.

**Review by sberyozkin — APPROVED** [review]
@lordofthejars thanks; lets wait for the feedback from @vsevel

**Comment by vsevel** [comment]
dear @lordofthejars & @sberyozkin,
this is some very nice work.
my main comment revolves around some exchanges we had earlier about this feature's applicability in https://github.com/quarkusio/quarkus/issues/7007#issuecomment-583395622
It is disabled by default and it is a good thing. I think we should stress in the documentation when to activate it, and/or the downsides of activating it.
People will have to realize that if they activate that feature, and vault goes rogue, the quarkus application will become not ready, will start not responding to external invocations, and most likely will trigger some monitoring alerts. If you have 200 microservices fetching secrets from vault, and vault gets down, all 200 microservices will turn red, whereas there is nothing wrong with those microservices. and it might take some time for ops to figure out that 200 incidents are just one.
To me when vault gets down, it should be red in the monitoring console, and all services depending on it should be orange to indicate that they are not running at 100% capacity.
So the vault healthcheck is really useful in situations where your microservice is dependent upon vault for 100% of its use cases (which I think is going to be rare). And still, making the service not ready is probably not going to make any difference for my clients since they are going to get a 503 instead of a 500. 
That said I am seeing other healthcheck impl for external systems (neo4j, mongo, agroal - but not for ES). So may be this is the norm?

To wrap up on this issue I would add some doc saying:
> activate this healthcheck in situations where you do not want your application to be reachable if vault is not available (for instance in situations where all your microservice's features are dependent on vault). 
> understand that if vault goes down, or switches to one of the statuses picked up by the healthcheck, all microservice replicas will become unreachable, as if they are not running.

Am I too dramatic? ;-)

other minor comments:
 - there is a bit of copy/pasting between `getBuildConfig()` and `getConfig()` in `VaultConfigSource`
 - I had already implemented the `sys/health` endpoint, but it was only used for testing purposes. so we located it into `io.quarkus.vault.test.client.dto` and used only from `TestVaultClient`, with other use cases such as `sys/init`, `sys/seal-status`, ... My version had a complete representation of the response, but is lacking the 2 optional arguments (`standbyok ` and `perfstandbyok`). it would be nice to fully implement the `sys/health` api (including the other arguments such as `activecode`, `standbycode`, ...), plus remove my old code from `TestVaultClient` and use the new code in `VaultTestExtension.initVault()`

what do you think?

**Comment by sberyozkin** [comment]
Hi @vsevel Thanks a million for the feedback, I'll check in detail a bit later, lets discuss until you and @lordofthejars agree.

**Comment by lordofthejars** [comment]
Ok, thins I am going to do:

. Resolve conflicts
. Update the sys API with the code that was developed for testing purposes
. Update docs with a warning
. Remove the Test client for doing _sys_ calls and use the new one

**Comment by lordofthejars** [comment]
Damn it Git. Now let me check how I fix this

**Comment by lordofthejars** [comment]
I close this PR and I open a new one with all this thing fixed. Sorry.

**Pull request closed without merging**
